### PR TITLE
Support retrieving blame information #798

### PR DESCRIPTION
**Pull request by panychek**
This patch allows to get blame data over [the file blame API](https://docs.gitlab.com/ce/api/repository_files.html#get-file-blame-from-repository).

### CODE REVIEW

File: repository_files.go [hunk 1]
```diff
@@ -142,6 +143,66 @@ func (s *RepositoryFilesService) GetFileMetaData(pid interface{}, fileName strin
 	return f, resp, err
 }
 
+// FileBlameRange represents one item of blame information.
+//
+// GitLab API docs: https://docs.gitlab.com/ce/api/repository_files.html#get-file-blame-from-repository
```
Comment (svanharmelen) [inline repository_files.go]:
```suggestion
// GitLab API docs: 
// https://docs.gitlab.com/ce/api/repository_files.html#get-file-blame-from-repository
```

File: repository_files.go [hunk 2]
```diff
@@ -142,6 +143,66 @@ func (s *RepositoryFilesService) GetFileMetaData(pid interface{}, fileName strin
...
+// GetFileBlameOptions represents the available GetFileBlame() options.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ce/api/repository_files.html#get-file-blame-from-repository
+type GetFileBlameOptions struct {
+	Ref *string `url:"ref,omitempty" json:"ref,omitempty"`
+}
+
+// GetFileBlame allows you to receive blame information.
+// Each blame range contains lines and corresponding commit info.
```
Comment (svanharmelen) [inline repository_files.go]:
```suggestion
// GetFileBlame allows you to receive blame information. Each blame range
// contains lines and corresponding commit info.
```

File: repository_files.go [hunk 3]
```diff
@@ -142,6 +143,66 @@ func (s *RepositoryFilesService) GetFileMetaData(pid interface{}, fileName strin
...
+		ParentIDs      []string   `json:"parent_ids"`
+		Message        string     `json:"message"`
+		AuthoredDate   *time.Time `json:"authored_date"`
+		AuthorName     string     `json:"author_name"`
+		AuthorEmail    string     `json:"author_email"`
+		CommittedDate  *time.Time `json:"committed_date"`
+		CommitterName  string     `json:"committer_name"`
+		CommitterEmail string     `json:"committer_email"`
+	} `json:"commit"`
+	Lines []string `json:"lines"`
```
Comment (svanharmelen) [inline repository_files.go]: We already have a `Commit` type, can't we reuse that one here?

File: repository_files.go [hunk 4]
```diff
@@ -142,6 +143,66 @@ func (s *RepositoryFilesService) GetFileMetaData(pid interface{}, fileName strin
 	return f, resp, err
 }
 
+// FileBlameRange represents one item of blame information.
+//
+// GitLab API docs: https://docs.gitlab.com/ce/api/repository_files.html#get-file-blame-from-repository
+type FileBlameRange struct {
```
Comment (svanharmelen) [inline repository_files.go]: I prefer to call this just `FileBlame` as it doesn't specify a range.

**Comment by panychek** [comment]
Yeah, thanks. Sorry, I didn't have time to fix it myself.

As to the naming, [the original GitLab entity](https://gitlab.com/gitlab-org/gitlab-foss/-/blob/3bb3ac3d53398a8e2e5d2de95ee2a226db84efe2/lib/api/entities.rb#L13) is named `BlameRange`, so I decided to stick with it.

**Pull request merged**